File: backend/app/providers/aerodatabox.py
from __future__ import annotations

from typing import Any

import httpx

from app.config import Settings
from app.models import BoardFlight


class AeroDataBoxProvider:
    name = "aerodatabox"

    def __init__(self, settings: Settings, client: httpx.AsyncClient):
        self._s = settings
        self._client = client

    async def board(self, iata: str, direction: str = "Departure") -> list[BoardFlight]:
        """Airport FIDS around now. Scheduled/estimated/actual — not a fare, not ADS-B."""
        r = await self._client.get(
            f"https://aerodatabox.p.rapidapi.com/flights/airports/iata/{iata}",
            params={
                "offsetMinutes": "-120",
                "durationMinutes": "720",
                "withLeg": "true",
                "direction": direction,
                "withCancelled": "true",
                "withCodeshared": "true",
                "withCargo": "false",
                "withPrivate": "false",
            },
            headers={
                "X-RapidAPI-Key": self._s.rapidapi_key,
                "X-RapidAPI-Host": "aerodatabox.p.rapidapi.com",
            },
            timeout=25.0,
        )
        if r.status_code >= 400:
            return []
        body = r.json()
        key = "departures" if direction.lower().startswith("dep") else "arrivals"
        flights: list[BoardFlight] = []
        for raw in body.get(key) or []:
            parsed = _one(raw, iata, direction)
            if parsed:
                flights.append(parsed)
        return flights[:40]


def _one(raw: dict[str, Any], iata: str, direction: str) -> BoardFlight | None:
    number = raw.get("number") or raw.get("callSign")
    if not number:
        return None
    airline = (raw.get("airline") or {}).get("iata") or (raw.get("airline") or {}).get("name")
    dep = raw.get("departure") or raw.get("movement") or {}
    arr = raw.get("arrival") or {}
    if direction.lower().startswith("dep"):
        other = (arr.get("airport") or {}).get("iata")
        sched = (dep.get("scheduledTime") or {}).get("local") or dep.get("scheduledTimeLocal")
        est = (dep.get("revisedTime") or {}).get("local") or dep.get("revisedTimeLocal")
        terminal = dep.get("terminal")
        gate = dep.get("gate")
        origin, dest = iata, other
    else:
        other = (dep.get("airport") or {}).get("iata")
        sched = (arr.get("scheduledTime") or {}).get("local") or arr.get("scheduledTimeLocal")
        est = (arr.get("revisedTime") or {}).get("local") or arr.get("revisedTimeLocal")
        terminal = arr.get("terminal")
        gate = arr.get("gate")
        origin, dest = other, iata
    status = (raw.get("status") or "") if isinstance(raw.get("status"), str) else str(raw.get("status") or "")
    return BoardFlight(
        flight_number=str(number),
        carrier=airline,
        origin=origin,
        dest=dest,
        scheduled=str(sched)[:19] if sched else None,
        estimated=str(est)[:19] if est else None,
        status=status or None,
        terminal=str(terminal) if terminal else None,
        gate=str(gate) if gate else None,
        source="aerodatabox",
        layer="schedule-status",
    )

File: backend/src/providers/aerodatabox.rs
use std::time::Duration;

use serde_json::Value;

use crate::config::Settings;
use crate::models::BoardFlight;

pub struct AeroDataBoxProvider {
    key: String,
    client: reqwest::Client,
}

impl AeroDataBoxProvider {
    pub fn new(settings: &Settings, client: reqwest::Client) -> Self {
        Self {
            key: settings.rapidapi_key.clone(),
            client,
        }
    }

    pub async fn board(&self, iata: &str, direction: &str) -> anyhow::Result<Vec<BoardFlight>> {
        let r = self
            .client
            .get(format!("https://aerodatabox.p.rapidapi.com/flights/airports/iata/{iata}"))
            .query(&[
                ("offsetMinutes", "-120"),
                ("durationMinutes", "720"),
                ("withLeg", "true"),
                ("direction", direction),
                ("withCancelled", "true"),
                ("withCodeshared", "true"),
                ("withCargo", "false"),
                ("withPrivate", "false"),
            ])
            .header("X-RapidAPI-Key", &self.key)
            .header("X-RapidAPI-Host", "aerodatabox.p.rapidapi.com")
            .timeout(Duration::from_secs(25))
            .send()
            .await?;
        if r.status().as_u16() >= 400 {
            return Ok(vec![]);
        }
        let body: Value = r.json().await.unwrap_or(serde_json::json!({}));
        let key = if direction.to_lowercase().starts_with("dep") {
            "departures"
        } else {
            "arrivals"
        };
        let mut flights = Vec::new();
        for raw in body.get(key).and_then(|v| v.as_array()).cloned().unwrap_or_default() {
            if let Some(parsed) = one(&raw, iata, direction) {
                flights.push(parsed);
            }
        }
        flights.truncate(40);
        Ok(flights)
    }
}

fn one(raw: &Value, iata: &str, direction: &str) -> Option<BoardFlight> {
    let number = raw
        .get("number")
        .or_else(|| raw.get("callSign"))
        .and_then(|v| v.as_str())?;
    let airline = raw
        .get("airline")
        .and_then(|a| a.get("iata").or_else(|| a.get("name")))
        .and_then(|v| v.as_str())
        .map(|s| s.to_string());
    let dep = raw.get("departure").or_else(|| raw.get("movement")).cloned().unwrap_or(serde_json::json!({}));
    let arr = raw.get("arrival").cloned().unwrap_or(serde_json::json!({}));
    let (origin, dest, sched, est, terminal, gate) = if direction.to_lowercase().starts_with("dep") {
        let other = arr.get("airport").and_then(|a| a.get("iata")).and_then(|v| v.as_str());
        let sched = dep
            .get("scheduledTime")
            .and_then(|t| t.get("local"))
            .or_else(|| dep.get("scheduledTimeLocal"));
        let est = dep
            .get("revisedTime")
            .and_then(|t| t.get("local"))
            .or_else(|| dep.get("revisedTimeLocal"));
        (
            Some(iata.to_string()),
            other.map(|s| s.to_string()),
            sched.and_then(|v| v.as_str()).map(|s| s.chars().take(19).collect()),
            est.and_then(|v| v.as_str()).map(|s| s.chars().take(19).collect()),
            dep.get("terminal").and_then(|v| v.as_str()).map(|s| s.to_string()),
            dep.get("gate").and_then(|v| v.as_str()).map(|s| s.to_string()),
        )
    } else {
        let other = dep.get("airport").and_then(|a| a.get("iata")).and_then(|v| v.as_str());
        let sched = arr
            .get("scheduledTime")
            .and_then(|t| t.get("local"))
            .or_else(|| arr.get("scheduledTimeLocal"));
        let est = arr
            .get("revisedTime")
            .and_then(|t| t.get("local"))
            .or_else(|| arr.get("revisedTimeLocal"));
        (
            other.map(|s| s.to_string()),
            Some(iata.to_string()),
            sched.and_then(|v| v.as_str()).map(|s| s.chars().take(19).collect()),
            est.and_then(|v| v.as_str()).map(|s| s.chars().take(19).collect()),
            arr.get("terminal").and_then(|v| v.as_str()).map(|s| s.to_string()),
            arr.get("gate").and_then(|v| v.as_str()).map(|s| s.to_string()),
        )
    };
    let status = raw.get("status").and_then(|v| {
        if v.is_string() {
            v.as_str().map(|s| s.to_string())
        } else {
            Some(v.to_string())
        }
    });
    Some(BoardFlight {
        flight_number: number.to_string(),
        carrier: airline,
        origin,
        dest,
        scheduled: sched,
        estimated: est,
        status: status.filter(|s| !s.is_empty()),
        terminal: terminal.filter(|s| !s.is_empty()),
        gate: gate.filter(|s| !s.is_empty()),
        source: "aerodatabox".into(),
        layer: "schedule-status".into(),
    })
}
